mito-ai/mito_ai/providers.py

```diff
@@ -398,32 +284,112 @@ async def stream_and_save_completions(
             model: The model to request completions for.
             message_id: The message ID to track the request.
+            reply_fn: Function to call with each chunk for streaming replies.
             response_format_info: Optional response format information.
-            reply_fn: Optional function to call with each chunk for streaming replies.
             
         Returns:
             The accumulated response string.
         """
-        # Create a request object for the streaming
-        request = CompletionRequest(
-            type=message_type,
-            message_id=message_id,
-            messages=messages,
-            stream=True
-        )
+        # Reset the last error
+        self.last_error = None
         
         # Use a string buffer to accumulate the full response
         accumulated_response = ""
         
-        # Stream completions 
-        async for reply in self.stream_completions(request, message_type, model, user_input):
-            
-            # Always call reply_fn if it exists, regardless of reply type
-            if reply_fn:
-                reply_fn(reply)
+        # Validate that the model is supported. If not fall back to gpt-4o-mini
+        if model not in self.models:
+            model = "gpt-4o-mini"
             
-            # Only accumulate content from actual completion chunks
-            if isinstance(reply, CompletionStreamChunk):
-                accumulated_response += reply.chunk.content
+        # Send initial acknowledgment
+        reply_fn(CompletionReply(
+            items=[
+                CompletionItem(content="", isIncomplete=True, token=message_id)
+            ],
+            parent_id=message_id,
+        ))
+        
+        # Get the last message content for logging
+        last_message_content = str(messages[-1].get("content", "")) if messages else ""
+        
+        # Prepare completion function parameters
+        completion_function_params = get_open_ai_completion_function_params(
+            model, messages, True, response_format_info
+        )
+        
+        # Stream completions based on the available client
+        if self._openAI_async_client is not None:
+            # Stream from OpenAI
+            try:
+                client = self._openAI_async_client
+                if client is None:
+                    raise ValueError("OpenAI client not initialized")
+                    
+                stream: AsyncStream[ChatCompletionChunk] = await client.chat.completions.create(**completion_function_params)
+                
+            except BaseException as e:
+                self.last_error = CompletionError.from_exception(e)
+                log(
+                    MITO_AI_COMPLETION_ERROR, 
+                    params={
+                        KEY_TYPE_PARAM: USER_KEY,
+                        'message_type': message_type.value,
+                    },
+                    error=e
+                )
+                raise
```

Comment:
Right now we have these two try, except blocks in this if statement. Is that necessary for error handling or can we just have one try block and one except block at the end of it?